src/LupaClient.php

```diff
@@ -0,0 +1,213 @@
+<?php
+
+namespace LupaSearch;
+
+use GuzzleHttp\Exception\ConnectException;
+use GuzzleHttp\Exception\GuzzleException;
+use LupaSearch\Exceptions\AuthenticationException;
+use LupaSearch\Exceptions\MissingCredentialsException;
+use LupaSearch\Exceptions\TooManyRetriesException;
+use LupaSearch\Factories\HttpClientFactory;
+use LupaSearch\Factories\HttpClientFactoryInterface;
+use LupaSearch\Utils\JwtUtils;
+use GuzzleHttp\Client;
+use GuzzleHttp\ClientInterface;
+use GuzzleHttp\Exception\ClientException;
+use GuzzleHttp\Psr7\Request;
+
+use function json_encode;
+
+class LupaClient
+{
+    const VERSION = '0.1.0';
+
+    const API_BASE_PATH = 'https://api.lupasearch.com/v1';
+    const USER_AGENT = 'LupaSearch API PHP Client, v' . self::VERSION;
+
+    const METHOD_GET = 'GET';
+    const METHOD_POST = 'POST';
+    const METHOD_PUT = 'PUT';
+    const METHOD_DELETE = 'DELETE';
+
+    const REQUEST_MAX_RETRIES = 3;
+
+    const DEFAULT_HEADERS = [
+        'Content-Type' => 'application/json',
+        'Accept' => 'application/json',
+        'User-Agent' => self::USER_AGENT
+    ];
+
+    /**
+     * @var HttpClientFactoryInterface|null
+     */
+    private $httpClientFactory;
+
+    /**
+     * @var ClientInterface|null
+     */
+    private $httpClient;
+
+    /**
+     * @var string|null
+     */
+    private $jwtToken;
+
+    /**
+     * @var string|null
+     */
+    private $email;
+
+    /**
+     * @var string|null
+     */
+    private $password;
+
+    public function __construct(
+        HttpClientFactoryInterface $httpClientFactory = null
+    ) {
+        $this->httpClientFactory =
+            $httpClientFactory ?? new HttpClientFactory();
+    }
+
+    private function getHttpClient(): Client
+    {
+        if (null === $this->httpClient) {
+            $this->setHttpClient($this->httpClientFactory->create());
+        }
+
+        return $this->httpClient;
+    }
+
+    private function setHttpClient(ClientInterface $httpClient): self
+    {
+        $this->httpClient = $httpClient;
+
+        return $this;
+    }
+
+    private function setJwtToken(?string $jwtToken): self
+    {
+        $this->jwtToken = $jwtToken;
+
+        return $this;
+    }
+
+    private function getJwtToken(): ?string
+    {
+        return $this->jwtToken;
+    }
+
+    public function setEmail(string $email): self
+    {
+        $this->email = $email;
```

Comment:
username

Reply:
Currently we have login by email and password